solidity/contracts/liquidity-protection/LiquidityProtectionStats.sol

```diff
@@ -0,0 +1,231 @@
+// SPDX-License-Identifier: SEE LICENSE IN LICENSE
+pragma solidity 0.6.12;
+
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "@openzeppelin/contracts/utils/EnumerableSet.sol";
+
+import "./interfaces/ILiquidityProtectionStats.sol";
+import "../utility/Utils.sol";
+import "../utility/SafeMath.sol";
+import "../token/interfaces/IDSToken.sol";
+import "../token/interfaces/IERC20Token.sol";
+
+/**
+ * @dev This contract aggregates the statistics of the liquidity protection mechanism.
+ */
+contract LiquidityProtectionStats is ILiquidityProtectionStats, AccessControl, Utils {
+    using EnumerableSet for EnumerableSet.AddressSet;
+    using SafeMath for uint256;
+
+    bytes32 public constant ROLE_SUPERVISOR = keccak256("ROLE_SUPERVISOR");
+    bytes32 public constant ROLE_SEEDER = keccak256("ROLE_SEEDER");
+    bytes32 public constant ROLE_OWNER = keccak256("ROLE_OWNER");
+
+    mapping(IDSToken => uint256) private _totalPoolAmount;
+    mapping(IDSToken => mapping(IERC20Token => uint256)) private _totalReserveAmount;
+    mapping(IDSToken => mapping(IERC20Token => mapping(address => uint256))) private _totalProviderAmount;
+
+    mapping(address => EnumerableSet.AddressSet) private _providerPools;
+
+    // allows execution only by an owner
+    modifier ownerOnly {
```

Comment:
still no binary size optimization here

Reply:
The binary size of this contract is not an issue, and optimizing it (very slightly) means increasing the gas cost during runtime (very slightly). Are you sure you want this?

Reply:
Done